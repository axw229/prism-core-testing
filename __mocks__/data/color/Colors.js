// @flow
import cloneDeep from 'lodash/cloneDeep'
import sample from 'lodash/sample'

import type { ColorSetPayload, Color } from '../../../src/shared/types/Colors'

export const Colors: ColorSetPayload = {
  "Red": [
    [
      {
        "colorNumber": "6561",
        "coordinatingColors": {
          "coord2ColorId": 1926,
          "coord1ColorId": 2250,
          "whiteColorId": 2681
        },
        "description": [
          "Subdued",
          "Muted",
          "Brilliant",
          "Radiant"
        ],
        "id": "2248",
        "isExterior": true,
        "isInterior": true,
        "name": "Teaberry",
        "rgb": 15454683,
        "lrv": 68.551,
        "colorFamilyNames": [
          "Red"
        ],
        "brandKey": "SW",
        "red": 235,
        "green": 209,
        "blue": 219,
        "hue": 0.935897435897436,
        "saturation": 0.39393939393939387,
        "lightness": 0.8705882352941177,
        "hex": "#ebd1db",
        "isDark": false,
        "cssrgb": "rgb(235,209,219)",
        "storeStripLocator": "102-C1",
        "similarColors": [
          "2255",
          "2262",
          "2269",
          "2662",
          "2535",
          "2276",
          "2658",
          "1974",
          "2543",
          "2531"
        ]
      },
      {
        "colorNumber": "6568",
        "coordinatingColors": {
          "coord2ColorId": 11244,
          "coord1ColorId": 2687,
          "whiteColorId": 2681
        },
        "description": [
          "Subdued",
          "Muted",
          "Brilliant",
          "Radiant"
        ],
        "id": "2255",
        "isExterior": true,
        "isInterior": true,
        "name": "Lighthearted Pink",
        "rgb": 15586781,
        "lrv": 70.758,
        "brandedCollectionIds": [
          31797,
          31791
        ],
        "brandedCollectionNames": [
          "Pottery Barn Fall/Winter 2018",
          "Pottery Barn Fall/Winter 2018"
        ],
        "colorFamilyNames": [
          "Red"
        ],
        "brandKey": "SW",
        "red": 237,
        "green": 213,
        "blue": 221,
        "hue": 0.9444444444444445,
        "saturation": 0.3999999999999999,
        "lightness": 0.8823529411764706,
        "hex": "#edd5dd",
        "isDark": false,
        "cssrgb": "rgb(237,213,221)",
        "storeStripLocator": "103-C1",
        "similarColors": [
          "2248",
          "2262",
          "2269",
          "2276",
          "1974",
          "1981",
          "2658",
          "1995",
          "2662",
          "1988"
        ]
      }
    ]
  ]
}

export function getColors (): ColorSetPayload {
  return cloneDeep(Colors)
}

export function getColor (): Color | void {
  const colors = getColors()
  const fam1 = Object.keys(colors)[0]
  if (colors[fam1]) {
    return sample(colors[fam1][0])
  }
}
