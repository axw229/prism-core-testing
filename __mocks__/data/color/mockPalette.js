const mockPalette = {
  'colors': [
    {
      'colorNumber': '6765',
      'coordinatingColors': {
        'coord2ColorId': '11194',
        'coord1ColorId': '2870',
        'whiteColorId': '2162'
      },
      'description': [
        'Fairly Colorful',
        'Moderately Colorful',
        'Pale'
      ],
      'id': '2452',
      'isExterior': true,
      'isInterior': true,
      'name': 'Spa',
      'lrv': 64.476,
      'brandedCollectionNames': [],
      'colorFamilyNames': [
        'Blue'
      ],
      'brandKey': 'SW',
      'red': 167,
      'green': 220,
      'blue': 220,
      'hue': 0.5,
      'saturation': 0.43089430894308944,
      'lightness': 0.7588235294117647,
      'hex': '#a7dcdc',
      'isDark': false,
      'storeStripLocator': '164-C2',
      'similarColors': [
        '2634',
        '2630',
        '11225',
        '2472',
        '2445',
        '2622',
        '2459',
        '1606',
        '11226',
        '2451'
      ],
      'ignore': false,
      'archived': false
    },
    {
      'colorNumber': '6860',
      'coordinatingColors': {
        'coord2ColorId': '11339',
        'coord1ColorId': '11364',
        'whiteColorId': '2681'
      },
      'description': [
        'Soft',
        'Fairly Bright'
      ],
      'id': '2547',
      'isExterior': true,
      'isInterior': true,
      'name': 'Eros Pink',
      'lrv': 18.874,
      'brandedCollectionNames': [
        '2020 Play',
        '2019 Naturalist'
      ],
      'colorFamilyNames': [
        'Red'
      ],
      'brandKey': 'SW',
      'red': 200,
      'green': 79,
      'blue': 104,
      'hue': 0.9655647382920111,
      'saturation': 0.5238095238095238,
      'lightness': 0.5470588235294118,
      'hex': '#c84f68',
      'isDark': false,
      'storeStripLocator': '105-C5',
      'similarColors': [
        '2273',
        '2266',
        '2542',
        '2546',
        '1998',
        '2280',
        '2530',
        '2287',
        '2279',
        '30193'
      ],
      'ignore': false,
      'archived': false
    },
    {
      'colorNumber': '9018',
      'coordinatingColors': {
        'coord2ColorId': '11218',
        'coord1ColorId': '11200',
        'whiteColorId': '2366'
      },
      'description': [
        'Intense',
        'Pale'
      ],
      'id': '11195',
      'isExterior': false,
      'isInterior': true,
      'name': 'Honey Bees',
      'lrv': 70.14,
      'brandedCollectionNames': [
        '2018 Unity'
      ],
      'colorFamilyNames': [
        'Yellow'
      ],
      'brandKey': 'SW',
      'red': 251,
      'green': 214,
      'blue': 130,
      'hue': 0.11570247933884299,
      'saturation': 0.937984496124031,
      'lightness': 0.7470588235294118,
      'hex': '#fbd682',
      'isDark': false,
      'storeStripLocator': '134-C4',
      'similarColors': [
        '2382',
        '11199',
        '11196',
        '2588',
        '2375',
        '11198',
        '11194',
        '11191',
        '3146',
        '2361'
      ],
      'ignore': false,
      'archived': false
    },
    {
      'colorNumber': '9007',
      'coordinatingColors': {
        'coord2ColorId': '1666',
        'coord1ColorId': '2686',
        'whiteColorId': '2331'
      },
      'description': [
        'Rich',
        'Vibrant',
        'Pale'
      ],
      'id': '11184',
      'isExterior': false,
      'isInterior': true,
      'name': 'Melón Meloso',
      'lrv': 55.314,
      'brandedCollectionNames': [],
      'colorFamilyNames': [
        'Orange'
      ],
      'brandKey': 'SW',
      'red': 242,
      'green': 184,
      'blue': 140,
      'hue': 0.0718954248366013,
      'saturation': 0.7968749999999999,
      'lightness': 0.7490196078431373,
      'hex': '#f2b88c',
      'isDark': false,
      'storeStripLocator': '121-C3',
      'similarColors': [
        '2334',
        '2983',
        '2327',
        '11188',
        '2326',
        '2045',
        '2038',
        '3012',
        '2341',
        '2031'
      ],
      'ignore': false,
      'archived': false
    }
  ],
  'activeColor': {
    'colorNumber': '9007',
    'coordinatingColors': {
      'coord2ColorId': '1666',
      'coord1ColorId': '2686',
      'whiteColorId': '2331'
    },
    'description': [
      'Rich',
      'Vibrant',
      'Pale'
    ],
    'id': '11184',
    'isExterior': false,
    'isInterior': true,
    'name': 'Melón Meloso',
    'lrv': 55.314,
    'brandedCollectionNames': [],
    'colorFamilyNames': [
      'Orange'
    ],
    'brandKey': 'SW',
    'red': 242,
    'green': 184,
    'blue': 140,
    'hue': 0.0718954248366013,
    'saturation': 0.7968749999999999,
    'lightness': 0.7490196078431373,
    'hex': '#f2b88c',
    'isDark': false,
    'storeStripLocator': '121-C3',
    'similarColors': [
      '2334',
      '2983',
      '2327',
      '11188',
      '2326',
      '2045',
      '2038',
      '3012',
      '2341',
      '2031'
    ],
    'ignore': false,
    'archived': false
  },
  'toggleCompareColor': false,
  'temporaryActiveColor': null
}

export default mockPalette
