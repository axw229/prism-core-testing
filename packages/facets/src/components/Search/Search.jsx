// @flow
import React, { useContext, useEffect, useState } from 'react'
import { FormattedMessage } from 'react-intl'
import { useSelector } from 'react-redux'
// $FlowIgnore -- no defs for react-virtualized
import { AutoSizer,Grid } from 'react-virtualized'
import Prism, { ColorSwatch } from '@prism/toolkit'
import * as GA from 'src/analytics/GoogleAnalytics'
import { colorSwatchCommonProps } from 'src/components/ColorSwatchContent/ColorSwatchContent'
import ColorWallContext, { type ColorWallContextProps } from 'src/components/Facets/ColorWall/ColorWallContext'
import { GA_TRACKER_NAME_BRAND } from 'src/constants/globals'
import ConfigurationContext, {
  type ConfigurationContextType
} from 'src/contexts/ConfigurationContext/ConfigurationContext'
import { fullColorNumber } from 'src/shared/helpers/ColorUtils'
import type { ColorsState } from 'src/shared/types/Actions'
import omitPrefix from 'src/shared/utils/omitPrefix.util'
import useEffectAfterMount from '../../shared/hooks/useEffectAfterMount'
import type { CrossSearch } from '../Facets/ColorSearchFacet/ColorSearchFacet'
import { SwatchContent } from '../Facets/ColorWall/ColorWallV3/Swatch/Swatch'
import TextButton from '../GeneralButtons/TextButton/TextButton'
import GenericMessage from '../Messages/GenericMessage'
import { GenericOverlay } from '../ToolkitComponents'
import './Search.scss'
import '../ColorSwatchContent/ColorSwatchContent.scss'
import 'src/scss/externalComponentSupport/AutoSizer.scss'

const baseClass = 'Search'
const EDGE_SIZE = 15

type SearchProps = {
  closeSearch?: () => void,
  contain?: boolean,
  crossSearch?: { query?: string, searching: boolean, onSearch: () => void } & CrossSearch,
  isChipLocator?: boolean
}

const Search = ({ closeSearch = () => {}, contain = false, crossSearch, isChipLocator }: SearchProps) => {
  const { results, count, suggestions, loading, query } = useSelector((state) => state.colors.search)
  const {
    items: { colorStatuses = {} }
  }: ColorsState = useSelector((state) => state.colors)
  const {
    addButtonText,
    colorDetailPageRoot,
    colorWallBgColor,
    colorWallPageRoot,
    displayDetailsLink,
    routeType
  }: ColorWallContextProps = useContext(ColorWallContext)
  const {
    brandId,
    brandKeyNumberSeparator,
    colorWall: { colorSwatch = {} }
  }: ConfigurationContextType = useContext(ConfigurationContext)
  const { colorNumOnBottom = false, houseShaped = false } = colorSwatch
  const [hasSearched, updateHasSearched] = useState(typeof count !== 'undefined')

  useEffectAfterMount(() => {
    updateHasSearched(true)
  }, [count, results, loading])

  useEffect(() => {
    if (!query.length) {
      updateHasSearched(false)
    }
  }, [query])

  const cellRenderer = ({ columnIndex, isScrolling, isVisible, key, parent, rowIndex, style }: any) => {
    const result = results && results[columnIndex + rowIndex * parent.props.columnCount]
    const swatchClass = houseShaped ? 'Search--house-shaped' : 'swatch-content'
    const _style = houseShaped
      ? { ...style, width: style.width - 20, padding: '10px', textAlign: 'start' }
      : { ...style, textAlign: 'start' }

    if (!result) return

    const isSwatchEnabled = colorStatuses[result.id]?.status !== 0

    return (
      <Prism key={key}>
        <ColorSwatch
          {...colorSwatchCommonProps({ brandKeyNumberSeparator, color: result })}
          className={swatchClass}
          flagged={!isSwatchEnabled}
          {...(isChipLocator
            ? {
                // lowes valspar/hgsw qr color search
                renderer: () => {
                  return (
                    <div className={`${baseClass}__chip-locator`}>
                      <div
                        className={`${swatchClass}__label swatch-content${
                          colorNumOnBottom ? '__name-number' : '__number-name'
                        }`}
                      >
                        <p className={`${swatchClass}__label--number`}>
                          {fullColorNumber(result.brandKey, result.colorNumber, brandKeyNumberSeparator)}
                        </p>
                        <p className={`${swatchClass}__label--name`}>{result.name}</p>
                      </div>
                      <div className={`${baseClass}__chip-locator--buttons`}>
                        <button
                          className={`${baseClass}__chip-locator--buttons__button ${result.isDark ? 'dark-color' : ''}`}
                          onClick={() => {
                            GA.event(
                              {
                                category: 'QR Color Wall Search',
                                action: 'Find Chip',
                                label: `${result.name} - ${result.colorNumber}`
                              },
                              GA_TRACKER_NAME_BRAND[brandId]
                            )
                            window.location.href =
                              crossSearch && crossSearch.searching
                                ? crossSearch.onClickFindChip(result)
                                : colorWallPageRoot?.(result)
                            closeSearch()
                          }}
                        >
                          Find Chip
                        </button>
                        <button
                          className={`${baseClass}__chip-locator--buttons__button ${result.isDark ? 'dark-color' : ''}`}
                          onClick={() => {
                            GA.event(
                              {
                                category: 'QR Color Wall Search',
                                action: 'View Color',
                                label: `${result.name} - ${result.colorNumber}`
                              },
                              GA_TRACKER_NAME_BRAND[brandId]
                            )
                            window.location.href =
                              crossSearch && crossSearch.searching
                                ? crossSearch.onClickViewColor(result)
                                : colorDetailPageRoot?.(result)
                          }}
                        >
                          View Color
                        </button>
                      </div>
                    </div>
                  )
                }
              } // colorchips color wall || canada
            : addButtonText || displayDetailsLink
            ? {
                renderer: () => {
                  return (
                    <SwatchContent
                      enabled={isSwatchEnabled}
                      message={colorStatuses[result.id]?.message}
                      color={result}
                      isOnlyUsedforSearch
                    />
                  )
                }
              }
            : null)}
          style={_style}
        />
      </Prism>
    )
  }

  return (
    <div className={baseClass}>
      <div className={`${baseClass}__results-pane`} style={{ backgroundColor: colorWallBgColor }}>
        {loading ? (
          <GenericOverlay type={GenericOverlay.TYPES.LOADING} semitransparent>
            <FormattedMessage id='SEARCH.SEARCHING' />
          </GenericOverlay>
        ) : !hasSearched ? (
          <GenericMessage type={GenericMessage.TYPES.NORMAL}>
            <FormattedMessage id='SEARCH.PROMPT' />
          </GenericMessage>
        ) : !count ? (
          <GenericMessage type={GenericMessage.TYPES.WARNING}>
            <FormattedMessage id='SEARCH.NO_RESULTS' />
            {suggestions && suggestions.length ? (
              <FormattedMessage
                id='SEARCH.SUGGESTIONS'
                values={{
                  suggestions: (
                    <>
                      {suggestions.map((suggestion, i, arr) => (
                        <React.Fragment key={i}>
                          <TextButton
                            className={routeType === 'memory' ? 'no-underline' : undefined}
                            to={routeType === 'memory' ? undefined : `./${omitPrefix(suggestion)}`}
                          >
                            {omitPrefix(suggestion)}
                          </TextButton>
                          {i < arr.length - 1 && ', '}
                        </React.Fragment>
                      ))}
                    </>
                  )
                }}
              />
            ) : null}
            {crossSearch && !crossSearch.searching ? (
              <strong>
                {crossSearch.text}{' '}
                <a
                  className={`${baseClass}__results-pane__subtitle--link`}
                  href={crossSearch.colorWallPageRoot}
                  target='_blank'
                >
                  Click Here
                </a>
              </strong>
            ) : null}
          </GenericMessage>
        ) : (
          <div
            className={`${baseClass}__results-pane__swatches ${
              contain ? `${baseClass}__results-pane__swatches--cover` : ''
            }`}
          >
            {crossSearch && (
              <p className={`${baseClass}__results-pane__subtitle`}>
                Looking for more colors?{' '}
                <a
                  className={`${baseClass}__results-pane__subtitle--link`}
                  href={crossSearch.colorWallPageRoot}
                  target='_blank'
                >
                  {`Search ${crossSearch.brand.name} colors.`}
                </a>
              </p>
            )}
            <AutoSizer disableHeight={!contain}>
              {({ height = 0, width }) => {
                const gridWidth = width - EDGE_SIZE * (houseShaped ? 0.7 : 2)
                const columnCount = Math.max(1, Math.round(gridWidth / 175))
                const newSize = gridWidth / columnCount
                const rowHeight = houseShaped ? 245 : newSize
                const rowCount = Math.ceil(results.length / columnCount)
                const gridHeight = contain ? height : Math.max(height, rowCount * newSize + EDGE_SIZE * 2)

                return (
                  <Grid
                    containerStyle={{
                      marginBotom: `${EDGE_SIZE}px`,
                      marginTop: `${EDGE_SIZE}px`
                    }}
                    cellRenderer={cellRenderer}
                    columnWidth={newSize}
                    columnCount={columnCount}
                    height={gridHeight}
                    rowHeight={rowHeight}
                    rowCount={rowCount}
                    width={width}
                  />
                )
              }}
            </AutoSizer>
          </div>
        )}
      </div>
    </div>
  )
}

export default Search
