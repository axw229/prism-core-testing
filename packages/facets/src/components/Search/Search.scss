.Search {
  font-family: var(getVarName('typography.titleFontFamily'));

  &__results-pane {
    align-items: center;
    display: flex;
    justify-content: center;
    min-height: 60vh;
    position: relative;
    text-align: center;

    @include respond-above(sm) { min-height: 66vh; }

    &__subtitle {
      color: #404040;
      margin: 1rem 1rem 0;
      text-align: left;
      line-height: 1.1rem;

      &--link {
        text-decoration: underline;
      }
    }

    &__swatches {
      align-self: flex-start;
      width: 100%;

      &--cover {
        height: 100%;
        overflow: hidden;
        position: absolute;
      }
    }
  }

  &__chip-locator {
    margin: .1rem;
    font-size: 1rem;
    line-height: 1.3rem;

    &--buttons {
      position: absolute;
      bottom: .8rem;
      width: 120px;

      &__button {
        width: 120px;
        margin-top: 10px !important;
        padding: 4px !important;
        border: 1px solid black;
        border-radius: 15px;
        font-size: inherit;
        font-weight: 500;
        line-height: 1.1rem;
      }

      .dark-color { border-color: white }
    }
  }

  &--house-shaped {
    height: 108px !important;
    margin-top: 48px !important;
    padding: 0 !important;
    box-shadow: none;
    outline: none;
    transition: none;

    &::before {
      content: '';
      display: block;
      position: absolute;
      top: -47%;
      width: 100%;
      height: 47.5%;
      background-color: inherit;
      clip-path: polygon(50% 0, 0 100%, 100% 100%);
      cursor: default;
    }
  }

  .ReactVirtualized__Grid__innerScrollContainer { margin: auto; }
}