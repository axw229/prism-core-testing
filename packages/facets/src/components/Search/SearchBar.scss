.SearchBar {
  $input-bg: var(getThemeColor('white'));
  $input-padding: 0.5em;
  $clear-size: 1em;
  font-family: var(getVarName('typography.titleFontFamily'));

  &__search-form {
    width: 100%;
  }

  &__label {
    cursor: default;
    display: inline-block;
    font-size: 0.9em;
    font-weight: bold;
    margin-bottom: 0.25em;
  }

  &__icon {
    margin: 0 $ui-spacing 0 0;
  }

  &__cancel-button {
    margin: 0 0 0 $ui-spacing;
  }

  &__inner {
    display: flex;
    flex-direction: row;
    align-items: center;
    padding: $ui-spacing;
  }

  &__wrapper {
    background: $input-bg;
    position: relative;
    margin: 0;
    flex: 1;
    height: $ui-height-min--mobile;
    border-radius: var(getVarName('typography.textInputBorderRadius'));

    @include respond-above(xs) {
      height: $ui-height-min;
    }

    &--without-outline {
      border: 1px solid $color-light-grey;
    }

    &--with-outline {
      border: 1px solid var(getThemeColor('primary'));
      box-shadow: inset 0 0 0 1px var(getThemeColor('primary'));
    }
  }

  &__input {
    outline: none;
    border: none;
    width: 100%;
    height: 100%;
    position: absolute;
    line-height: $ui-height-min--mobile;
    text-indent: $input-padding;
    padding: 0 $clear-size + (2 * $input-padding) 0 0;

    @include respond-above(xs) {
      line-height: $ui-height-min;
    }
  }

  .SearchBar__input::placeholder {
    color: var(getThemeColor('primaryPlaceholder'));
  }

  @media (max-width: 420px) {
    .SearchBar__input::placeholder {
      font-size: 1em;
    }
  }

  @media (max-width: 380px) {
    .SearchBar__input::placeholder {
      font-size: 0.8em;
    }
  }

  &__clean {
    position: absolute;
    right: ($clear-size / 2) + $input-padding;
    top: 50%;
    transform: translateY(-50%) translateX(50%);
    font-size: $clear-size;
    width: $clear-size * 1.5;
    height: $clear-size * 1.5;
    border-radius: 50%;

    &:hover,
    &:focus {
      box-shadow: inset 999em 0 0 0 rgba(white, 0.9);
      background: var(getThemeColor('primary'));
      color: var(getThemeColor('dark'));
    }
  }
}

.SearchBarLight {
  &__wrapper {
    border: none;
    color: #404040;
  }

  &__input::placeholder {
    color: #404040;
    font-size: inherit !important;
  }

  &__back {
    font-size: 1rem;
    margin-right: $ui-spacing;
  }

  &__icon {
    font-size: 12px;
    margin: 0 0 0 $ui-spacing;
    order: 3;
  }
}

.SearchBarMinimal {
  .button-bar__options__option__btn {
    background: none !important;
    color: black !important;
    text-transform: none;
  }

  .button-bar__options__option__btn,
  .SearchBar__input {
    font-size: 17px;
    font-weight: 600;
    line-height: 24px;
  }

  .button-bar__option-container {
    border: none !important;
  }
}
