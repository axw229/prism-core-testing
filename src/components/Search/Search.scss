.Search {
  &__search-form {
    display: flex;
    flex-direction: row;
    align-items: center;
    padding-bottom: $ui-spacing;
  }

  &__results-pane {
    background: #eee;
    min-height: 60vh;
    text-align: center;
    display: flex;
    align-items: center;
    justify-content: center;
    position: relative;

    @include respond-above(sm) {
      min-height: 60vw;
    }

    @include respond-above(md) {
      min-height: 50vw;
    }
  }
}
