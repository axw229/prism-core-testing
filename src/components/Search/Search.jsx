// @flow
import React, { PureComponent } from 'react'
import { connect } from 'react-redux'
// $FlowIgnore -- no defs for react-virtualized
import { Grid, AutoSizer } from 'react-virtualized'
import debounce from 'lodash/debounce'

import { loadSearchResults, clearSearch } from '../../store/actions/loadSearchResults'
import { add } from '../../store/actions/live-palette'

import { FormattedMessage } from 'react-intl'

import { generateColorDetailsPageUrl } from '../../shared/helpers/ColorUtils'

import ColorWallSwatch from '../Facets/ColorWall/ColorWallSwatch/ColorWallSwatch'
import SearchBar from './SearchBar'
import ButtonBar from '../ButtonBar/ButtonBar'
import GenericMessage from '../Messages/GenericMessage'
import CircleLoader from '../Loaders/CircleLoader/CircleLoader'

import { type Color } from '../../shared/types/Colors'

import './Search.scss'

type Props = {
  colors: void | any[],
  loading: Boolean,
  loadSearchResults: Function,
  clearSearch: Function,
  addToLivePalette: Function,
  onCancel: Function
}

type State = {
  resultSwatchSize: number // TODO: watch for resize and adjust this accordingly to maintain desired range of sizes
}

const SEARCH_DELAY = 500
const baseClass = 'Search'

export class Search extends PureComponent<Props, State> {
  state = {
    resultSwatchSize: 175
  }

  searchComponent: RefObject

  constructor (props: Props) {
    super(props)

    this.searchComponent = React.createRef()
    this.doClearSearch = this.doClearSearch.bind(this)
    this.cellRenderer = this.cellRenderer.bind(this)
  }

  render () {
    const { colors, onCancel, loading } = this.props
    const { resultSwatchSize } = this.state

    return (
      <div className={baseClass}>
        <form onSubmit={this.handleSubmit} className={`${baseClass}__search-form`}>
          <SearchBar onSearchInput={this.performSearch} onClearSearch={this.doClearSearch} ref={this.searchComponent} />
          <ButtonBar.Bar>
            <ButtonBar.Button onClick={onCancel}>
              <FormattedMessage id='CANCEL' />
            </ButtonBar.Button>
          </ButtonBar.Bar>
        </form>
        <div className={`${baseClass}__results-pane`}>
          { loading ? (
            <CircleLoader />
          ) : !colors ? (
            <GenericMessage type={GenericMessage.TYPES.NORMAL}>
              <FormattedMessage id='SEARCH_PROMPT' />
            </GenericMessage>
          ) : !colors.length ? (
            <GenericMessage type={GenericMessage.TYPES.WARNING}>
              <FormattedMessage id='SEARCH_NO_RESULTS' />
            </GenericMessage>
          ) : (
            <section className='color-wall-swatch-list color-wall-swatch-list--show-all'>
              <AutoSizer>
                {({ height, width }) => {
                  const columnCount = Math.round(width / resultSwatchSize)
                  const rowCount = Math.ceil(colors.length / columnCount)
                  const newSize = width / columnCount
                  return (
                    <Grid
                      colors={colors}
                      cellRenderer={this.cellRenderer}
                      columnWidth={newSize}
                      columnCount={columnCount}
                      height={height}
                      rowHeight={newSize}
                      rowCount={rowCount}
                      width={width}
                    />
                  )
                }}
              </AutoSizer>
            </section>
          ) }
        </div>
      </div>
    )
  }

  componentDidMount () {
    // clear search results whenever this component mounts -- this is how we control our initial "enter a color name" state
    this.props.clearSearch()
  }

  performSearch = debounce((value: string) => {
    this.props.loadSearchResults(value)
  }, SEARCH_DELAY)

  handleSubmit = (e: SyntheticInputEvent<HTMLInputElement>) => {
    e.preventDefault()
  }

  doClearSearch = () => {
    this.props.clearSearch()

    // if our search component has a focus method per its exposed API...
    if (typeof this.searchComponent.current[SearchBar.API.focus] === 'function') {
      // ... call it after clearing search
      this.searchComponent.current[SearchBar.API.focus]()
    }
  }

  cellRenderer = function cellRenderer ({
    columnIndex, // Horizontal (column) index of cell
    isScrolling, // The Grid is currently being scrolled
    isVisible, // This cell is visible within the grid (eg it is not an overscanned cell)
    key, // Unique key within array of cells
    parent, // Reference to the parent Grid (instance)
    rowIndex, // Vertical (row) index of cell
    style // Style object to be applied to cell (to position it)
  }: Object) {
    const { colors, addToLivePalette } = this.props
    const { props: { columnCount } } = parent

    const index = columnIndex + (rowIndex * columnCount)

    if (colors && colors[index]) {
      const thisColor: Color = colors[index]
      const linkToDetails: string = generateColorDetailsPageUrl(thisColor)

      return (
        <div key={key} style={style}>
          <ColorWallSwatch
            key={thisColor.hex}
            showContents
            color={thisColor}
            onAdd={addToLivePalette}
            detailsLink={linkToDetails}
          />
        </div>
      )
    }

    return null
  }
}

const mapStateToProps = (state, props) => {
  return {
    colors: state.colors.search.results,
    loading: state.colors.search.loading
  }
}

const mapDispatchToProps = (dispatch: Function) => {
  return {
    clearSearch: () => {
      dispatch(clearSearch())
    },
    loadSearchResults: (family) => {
      dispatch(loadSearchResults(family))
    },
    addToLivePalette: (color) => {
      dispatch(add(color))
    }
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Search)
