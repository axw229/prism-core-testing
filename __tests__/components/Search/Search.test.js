/* eslint-disable jest/no-disabled-tests */
/* eslint-env jest */

import React from 'react'
import { shallow } from 'enzyme'
import { Search } from 'src/components/Search/Search'
import * as Colors from '__mocks__/data/color/Colors'

const colors = Colors.getAllColors()

const mockFnLoadSearchResults = jest.fn()
const mockAddToLivePalette = jest.fn()
const mockPreventDefault = jest.fn()
const mockHandleSubmit = jest.fn()
const mockClearSearch = jest.fn()

const getSearch = (props) => {
  let defaultProps = {
    colors: [],
    loadSearchResults: mockFnLoadSearchResults,
    addToLivePalette: mockAddToLivePalette,
    clearSearch: mockClearSearch,
    loading: false
  }

  let newProps = Object.assign({}, defaultProps, props)
  return shallow(<Search {...newProps} />)
}

describe('Search with empty colors prop', () => {
  let search
  beforeEach(() => {
    if (!search) {
      search = getSearch()
    }
  })

  it('should match snapshot', () => {
    expect(search).toMatchSnapshot()
  })

  it('should render form', () => {
    expect(search.find('form').exists()).toBe(true)
  })

  it('should render SearchBar component', () => {
    expect(search.find('SearchBar').exists()).toBe(true)
  })

  it('should have state resultSwatchSize to equal 175', () => {
    expect(search.state('resultSwatchSize')).toEqual(175)
  })
})

describe('Search with colors prop', () => {
  let search
  beforeEach(() => {
    if (!search) {
      search = getSearch({ colors: colors })
    }
  })

  it('should match snapshot', () => {
    expect(search).toMatchSnapshot()
  })
})

describe('Search with events', () => {
  let search
  beforeEach(() => {
    if (!search) {
      search = getSearch()
    }
  })

  it('should call handleSubmit on form submit', () => {
    search.instance().handleSubmit = mockHandleSubmit
    search.instance().forceUpdate()
    search.find('form').simulate('submit', { preventDefault: mockPreventDefault })
    expect(mockHandleSubmit).toHaveBeenCalled()
  })
})
