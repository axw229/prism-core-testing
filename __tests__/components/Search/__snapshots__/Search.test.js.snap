// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search with colors prop should match snapshot 1`] = `
<div
  className="Search"
>
  <form
    className="Search__search-form"
    onSubmit={[Function]}
  >
    <SearchBar
      onClearSearch={[Function]}
      onSearchInput={[Function]}
    />
    <Bar>
      <Button>
        <FormattedMessage
          id="CANCEL"
          values={Object {}}
        />
      </Button>
    </Bar>
  </form>
  <div
    className="Search__results-pane"
  >
    <GenericMessage
      type="WARNING"
    >
      <FormattedMessage
        id="SEARCH_NO_RESULTS"
        values={Object {}}
      />
    </GenericMessage>
  </div>
</div>
`;

exports[`Search with empty colors prop should match snapshot 1`] = `
<div
  className="Search"
>
  <form
    className="Search__search-form"
    onSubmit={[Function]}
  >
    <SearchBar
      onClearSearch={[Function]}
      onSearchInput={[Function]}
    />
    <Bar>
      <Button>
        <FormattedMessage
          id="CANCEL"
          values={Object {}}
        />
      </Button>
    </Bar>
  </form>
  <div
    className="Search__results-pane"
  >
    <GenericMessage
      type="WARNING"
    >
      <FormattedMessage
        id="SEARCH_NO_RESULTS"
        values={Object {}}
      />
    </GenericMessage>
  </div>
</div>
`;
